File: 2024/day7.hs
import Control.Parallel.Strategies (parList, rdeepseq, using)
import Data.List (foldl')
import Data.Maybe (fromMaybe)
import System.IO (readFile)
import Text.Read (readMaybe)
import System.Clock (Clock(Monotonic), getTime, diffTimeSpec, toNanoSecs)

-- Evaluate recursively whether a combination of operations can produce the target
evaluateRecursive :: [Int] -> Int -> Int -> Bool
evaluateRecursive [] target current = current == target
evaluateRecursive (x:xs) target current =
  or [ evaluateRecursive xs target (current + x)
     , evaluateRecursive xs target (current * x)
     , if current /= 0 then evaluateRecursive xs target (read (show current ++ show x) :: Int) else False
     ]

-- Process a single line to determine if it matches the target
processLine :: String -> Int
processLine line =
  case words (map replaceColonWithSpace line) of
    (targetStr:rest) ->
      let target = read targetStr :: Int
          numbers = map read rest
       in if evaluateRecursive numbers target 0 then target else 0
    _ -> 0
  where
    replaceColonWithSpace ':' = ' '
    replaceColonWithSpace c = c

-- Calculate the total calibration result
calculateTotalCalibration :: [String] -> Int
calculateTotalCalibration lines =
  sum results
  where
    results = map processLine lines `using` parList rdeepseq

-- Main function

main :: IO ()
main = do
  start <- getTime Monotonic
  contents <- readFile "in.txt"
  let inputLines = filter (not . null) (lines contents)
  let totalCalibration = calculateTotalCalibration inputLines
  print $ "Total Calibration Result: " ++ show totalCalibration
  end <- getTime Monotonic
  let duration = fromIntegral (toNanoSecs (diffTimeSpec start end)) / 1e6
  putStrLn $ "Execution Time: " ++ show duration ++ "ms"

File: 2024/day7.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::time::Instant;

fn evaluate_recursive(numbers: &[i64], target: i64, current: i64) -> bool {
    if numbers.is_empty() {
        return current == target;
    }

    let next_number = numbers[0];
    let remaining = &numbers[1..];

    // Early termination
    if current > target && numbers.iter().all(|&n| n >= 0) {
        return false;
    }

    // Try addition
    if evaluate_recursive(remaining, target, current + next_number) {
        return true;
    }

    // Try multiplication
    if evaluate_recursive(remaining, target, current * next_number) {
        return true;
    }

    // Try concatenation
    if current != 0 {
        let concatenated = format!("{}{}", current, next_number)
            .parse::<i64>()
            .unwrap_or(0); // Handle potential overflow
        if evaluate_recursive(remaining, target, concatenated) {
            return true;
        }
    }

    false
}

fn process_line(line: &str) -> i64 {
    let parts: Vec<&str> = line.split(": ").collect();
    let target = parts[0].parse::<i64>().unwrap();
    let numbers: Vec<i64> = parts[1].split_whitespace().map(|x| x.parse().unwrap()).collect();

    if evaluate_recursive(&numbers, target, 0) {
        target
    } else {
        0
    }
}

fn calculate_total_calibration(input_lines: Vec<String>) -> i64 {
    input_lines
        .into_iter()
        .map(|line| process_line(&line))
        .sum()
}

fn main() {
    let start_time = Instant::now();
    let input_file = File::open("in.txt").expect("Failed to open file");
    let reader = io::BufReader::new(input_file);
    let input_lines: Vec<String> = reader.lines().map(|line| line.unwrap()).collect();

    let total_calibration = calculate_total_calibration(input_lines);
    let duration = start_time.elapsed();

    println!("Total Calibration Result: {}", total_calibration);
    println!("Execution Time: {:?}", duration);
}
